overcooked_ai_py/mdp/overcooked_mdp.py

```diff
@@ -669,27 +696,151 @@ def get_state_transition(self, state, joint_action):
         self.resolve_movement(new_state, joint_action)
 
         # Finally, environment effects
-        sparse_reward += self.step_environment_effects(new_state)
+        self.step_environment_effects(new_state)
 
         # Additional dense reward logic
         # shaped_reward += self.calculate_distance_based_shaped_reward(state, new_state)
 
-        return new_state, sparse_reward, shaped_reward
+        return new_state, sparse_reward, shaped_reward, events_infos
+
+    def get_empty_pots(self, pot_states):
+        return pot_states["empty"]
+
+    def get_ready_pots(self, pot_states):
+        return pot_states["tomato"]["ready"] + pot_states["onion"]["ready"]
 
-    def resolve_interacts(self, new_state, joint_action):
+    def get_cooking_pots(self, pot_states):
+        return pot_states["tomato"]["cooking"] + pot_states["onion"]["cooking"]
+
+    def get_full_pots(self, pot_states):
+        return self.get_cooking_pots(pot_states) + self.get_ready_pots(pot_states)
+
+    def get_partially_full_pots(self, pot_states):
+        return pot_states["tomato"]["partially_full"] + pot_states["onion"]["partially_full"]
+
+    def is_dish_pickup_useful(self, state, pot_states):
+        """
+        NOTE: this only works if self.num_players == 2
+        Useful if:
+        - Pot is ready/cooking and there is no player with a dish               \ 
+        - 2 pots are ready/cooking and there is one player with a dish          | -> number of dishes in players hands < number of ready/cooking/partially full soups 
+        - Partially full pot is ok if the other player is on course to fill it  /
+
+        We also want to prevent picking up and dropping dishes, so add the condition
+        that there must be no dishes on counters
+        """
+        # This next line is to prevent reward hacking (this logic is also used by reward shaping)
+        dishes_on_counters = self.get_counter_objects_dict(state)["dish"]
+        no_dishes_on_counters = len(dishes_on_counters) == 0
+
+        num_player_dishes = len(state.player_objects_by_type['dish'])
+        non_empty_pots = len(self.get_ready_pots(pot_states) + self.get_cooking_pots(pot_states) + self.get_partially_full_pots(pot_states))
+        return no_dishes_on_counters and num_player_dishes < non_empty_pots
+
+    def is_dish_drop_useful(self, state, pot_states, player_index):
+        """
+        NOTE: this only works if self.num_players == 2
+        Useful if:
+        - Onion is needed (all pots are non-full)
+        - Nobody is holding onions
+        """
+        if self.num_players != 2: return False
```

Comment:
Shouldn't this give an error if there aren't 2 players, rather than returning False? (I.e. "assert self.num_players == 2"

Reply:
I think the point is to allow the code to run with > 2 players, but just not collect these statistics in those cases.

Reply:
Yup, Rohin is right